p2p/node.go

```diff
@@ -287,6 +306,63 @@ func (n *Node) Connect(peerInfo peer.AddrInfo, timeout time.Duration) error {
 	return nil
 }
 
+// ProtectIP permanently adds the IP address of the given Multiaddr to a
+// list of protected IP addresses. Protected IPs can never be banned and will
+// not be added to the blacklist. If the IP address is already on the blacklist,
+// it will be removed.
+func (n *Node) ProtectIP(maddr ma.Multiaddr) error {
+	n.protectedIPsMut.Lock()
+	defer n.protectedIPsMut.Unlock()
+	ipNet, err := ipNetFromMaddr(maddr)
+	if err != nil {
+		return err
+	}
+	n.unbanIPNet(ipNet)
+	n.protectedIPs.Add(ipNet.IP.String())
+	return nil
+}
+
+// BanIP adds the IP address of the given Multiaddr to the blacklist. The
+// node will no longer dial or accept connections from this IP address. However,
+// if the IP address is protected, calling BanIP is a no-op. BanIP does not
+// automatically disconnect from the given multiaddress if there is currently an
+// open connection.
+func (n *Node) BanIP(maddr ma.Multiaddr) error {
+	ipNet, err := ipNetFromMaddr(maddr)
```

Comment:
So we return an error if the `multiAddr` doesn't include an IP address? I guess the caller would then check for that specific error and then noop for now?

Reply:
That's pretty much right. But there are at least some cases where the caller should not no-op. It's potentially a big problem if you try to ban a misbehaving IP address and it doesn't work. In any case it feels dangerous to silently fail here.